app/assets/v2/js/cart-ethereum-polygon.js

```diff
@@ -0,0 +1,430 @@
+    async estimateGasCost() {
+      // The below heuristics are used instead of `estimateGas()` so we can send the donation
+      // transaction before the approval txs are confirmed, because if the approval txs
+      // are not confirmed then estimateGas will fail.
+      let networkId = appCart.$refs.cart.networkId;
+
+      if (networkId !== '80001' && networkId !== '137' && appCart.$refs.cart.chainId !== '1') {
+        return;
+      }
+
+      // If we have a cart where all donations are in Dai, we use a linear regression to
+      // estimate gas costs based on real checkout transaction data, and add a 50% margin
+      const donationCurrencies = this.donationInputs.map(donation => donation.token);
+      const daiAddress = this.getTokenByName('DAI')?.addr;
+      const isAllDai = donationCurrencies.every((addr) => addr === daiAddress);
+
+      if (isAllDai) {
+        if (donationCurrencies.length === 1) {
+          // Special case since we overestimate here otherwise
+          return 70000;
+        }
+        // TODO: find a suitable curve using
+        // https://github.com/mds1/Gitcoin-Checkout-Gas-Analysis
+        // return 27500 * donationCurrencies.length + 125000;
+        return 10000 * donationCurrencies.length + 70000;
+      }
+
+      // Otherwise, based on contract tests, we use the more conservative heuristic below to get
+      // a gas estimate. The estimates used here are based on testing the cost of a single
+      // donation (i.e. one item in the cart). Because gas prices go down with batched
+      // transactions, whereas this assumes they're constant, this gives us a conservative estimate
+      const gasLimit = this.donationInputs.reduce((accumulator, currentValue) => {
+        const tokenAddr = currentValue.token?.toLowerCase();
+
+        if (currentValue.token === MATIC_ADDRESS) {
+          return accumulator + 70000; // MATIC donation gas estimate
+
+        } else if (tokenAddr === '0x960b236A07cf122663c4303350609A66A7B288C0'.toLowerCase()) {
+          return accumulator + 170000; // ANT donation gas estimate
+        }
+
+        return accumulator + 70000; // generic token donation gas estimate
+      }, 0);
+
+      return gasLimit;
+    },
+
+    // Returns true if user has enough balance within Polygon to avoid L1 deposit, false otherwise
+    async hasEnoughBalanceInPolygon() {
+      const requiredAmounts = {}; // keys are token symbols, values are required amounts as BigNumber
+
+      // Get total amount needed for eack token by summing over donation inputs
+      this.donationInputs.forEach((donation) => {
+        const tokenSymbol = donation.name;
+        const amount = toBigNumber(donation.amount);
+
+        if (!requiredAmounts[tokenSymbol]) {
+          // First time seeing this token, set the field and initial value
+          requiredAmounts[tokenSymbol] = { amount };
+        } else {
+          // We've seen this token, so just update the total
```

Comment:
i don't get this 😓 
> We've seen this token, so just update the total
maybe detail it a bit

Reply:
haha... i'll reword it